### PR TITLE
Implemented KNN algorithm using C++ (Closes issue #1789)

### LINKED ISSUE #1789: Add KNN Algorithm in C++
I am a GSSOC 2020 participant and I would love to implement the K Nearest Neighbors Algorithm using C++ as a part of GSSOC 2020. Please assign it to me @raghav-dalmia @asha15 @singh-shreya6

### PR DESCRIPTION
**Pull request by Aman-Codes**
Fixes #1789

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:
- Added the KNN Algorithm implemented using C++ to the Machine_Learning directory.
- Added Readme.md

#### Languages Used:
- C++

#### Files Added:
- K_Nearest_Neighbors.cpp
- Readme.md

@raghav-dalmia @asha15 @singh-shreya6

### CODE REVIEW

**Comment by asha15** [comment]
Remove the readme file, which is not relevant to this issue number

**Comment by Aman-Codes** [comment]
Done sir

**Comment by Aman-Codes** [comment]
@jainaman224 @asha15 @raghav-dalmia  @singh-shreya6 
Please review it.

File: Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
+#include <bits/stdc++.h>
+using namespace std;
+
+struct point {
+	float w,x,y,z; // w,x,y and z are the values of the dataset
```
Comment (jainaman224) [inline Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: Use 4 space indentation.

Comment (jainaman224) [reply Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: Add space after `,`.

File: Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	float distance; // distance will store the distance of dataset point from the unknown test point
+	int op; // op will store the class of the point
+};
+
+int main(){
```
Comment (jainaman224) [inline Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: Add space before `{`.

File: Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	}
+	cout << "\nw\tx\ty\tz\top\n";
+	for (i=0;i<n;i++){
+		cout << fixed << setprecision(2) << arr[i].w << '\t' << arr[i].x << '\t' << arr[i].y << '\t' << arr[i].z << '\t' << arr[i].op <<'\n';
+	}
+
+	cout << "\nEnter the feature values of the unknown point: \nw\tx\ty\tz\n";
+	cin >> p.w >> p.x >> p.y >> p.z;
+
+//Measuring the Euclidean distance
```
Comment (jainaman224) [inline Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: Inconsistent indentation.

**Comment by Aman-Codes** [comment]
@jainaman224 @asha15 @raghav-dalmia @singh-shreya6
Done the aforesaid changes. Kindly review it.

File: Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
+#include <bits/stdc++.h>
```
Comment (jainaman224) [inline Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: You have added this to Linear_Regression folder. Which is incorrect. Move this out of this folder.

File: Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp (hunk 1 above)
Comment (jainaman224) [reply Machine_Learning/Linear_Regression/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: Not fixed. I am not reviewing it further.

**Comment by Aman-Codes** [comment]
@jainaman224 @asha15 @raghav-dalmia @singh-shreya6
Did the asked changes. Kindly review it.

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+int main() {
+    int i, j, n, k;
+    struct point p;
+    struct point temp;
+    cout << "Number of data points: ";
+    cin >> n;
+    struct point arr[n];
+    cout << "\nEnter the dataset values: \n";
+    cout << "w\tx\ty\tz\top\n";
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    // Taking dataset
+    for (i = 0; i < n; i++) {
+        cin >> arr[i].w >> arr[i].x >> arr[i].y >> arr[i].z >> arr[i].op;
+    }
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    cout << "\nw\tx\ty\tz\top\n";
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    for (i = 0; i < n; i++) {
+        cout << fixed << setprecision(2) << arr[i].w << '\t' << arr[i].x << '\t' << arr[i].y << '\t' << arr[i].z << '\t' << arr[i].op << '\n';
+    }
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    cout << "\nEnter the feature values of the unknown point: \nw\tx\ty\tz\n";
+    cin >> p.w >> p.x >> p.y >> p.z;
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    // Measuring the Euclidean distance
+    for (i = 0; i < n; i++) {
+        arr[i].distance = sqrt(((arr[i].w - p.w) * (arr[i].w - p.w)) + ((arr[i].x - p.x) * (arr[i].x - p.x)) + ((arr[i].y - p.y) * (arr[i].y - p.y)) + ((arr[i].z - p.z) * (arr[i].z - p.z)));
+    }
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    // Sorting the training data with respect to distance
+    for (i = 1; i < n; i++) {
+        for (j = 0; j < n - i; j++) {
+            if (arr[j].distance > arr[j + 1].distance) {
+                temp = arr[j];
+                arr[j] = arr[j + 1];
+                arr[j + 1] = temp;
+            }
+        }
+    }
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    cout << "\nw\tx\ty\tz\top\tdistance\n";
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    for (i = 0; i < n; i++) {
+        cout << fixed << setprecision(2) << arr[i].w << '\t' << arr[i].x << '\t' << arr[i].y << '\t' << arr[i].z << '\t' << arr[i].op << '\t' << arr[i].distance << '\n';
+    }
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    // Taking the K nearest neighbors
+    cout << "\nNumber of nearest neighbors(k): ";
+    cin >> k;
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    int freq[1000] = {0};
+    int index = 0, maxfreq = 0;
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    // Creating frequency array of the class of k nearest neighbors
+    for (int i = 0; i < k; i++) {
+        freq[arr[i].op]++;
+    }
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

File: Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    // Finding the most frequent occurring class
+    for (int i = 0; i < 1000; i++) {
+        if(freq[i] > maxfreq) {
+            maxfreq = freq[i];
+            index = i;
+        }
+    }
```
Comment (asha15) [inline Machine_Learning/K_Nearest_Neighbors/K_Nearest_Neighbors.cpp]: keep a line space after this

**Comment by Aman-Codes** [comment]
@jainaman224 @asha15 @raghav-dalmia @singh-shreya6
Did the asked changes. Kindly review it.

**Comment by somya-kapoor** [comment]
@Aman-Codes Please squash the commits and then it will be merged.

**Comment by Aman-Codes** [comment]
@somya-kapoor @jainaman224 @asha15 
I have created a new pull request #2546 please check.

**Comment by somya-kapoor** [comment]
@Aman-Codes Please squash the commits

**Comment by Aman-Codes** [comment]
Can u please help me that how to squash commits. I am a beginner so I am having problem in squashing. @somya-kapoor

**Comment by Aman-Codes** [comment]
@somya-kapoor @jainaman224 @asha15 
It is showing showing some error in squashing so I have opened a new pull request #2546 with the final file and I am closing this pull request.

**Pull request closed without merging**